src/index.js

```diff
@@ -52,6 +53,7 @@ function makeShouldSkip () {
 function makeVisitor ({types: t}) {
   const shouldSkip = makeShouldSkip()
   return {
+    inherits: babelSyntaxObjectRestSpread,
```

Comment:
This should really be reverted. If something needs to support object rest spread, the user should be the one opting into it, it shouldn't be a weird side-effect of loading istanbul. If Jest specifically wants to auto-enable it, this should be loaded by Jest.

Reply:
@loganfsmyth @SimenB I'm keeping a close eye on https://github.com/babel/babel/issues/7660 -- I agree it feels a little weird (and like a slippery slope) in hindsight to be enabling one specific language feature.

Getting state-4 language features by default sounds like the best possible scenario (is this already the case?).

Reply:
> Getting state-4 language features by default sounds like the best possible scenario

No disagreement there. We've mostly been focused on 7, but there's probably a good argument to do a last pass on Babel 6 to enable stabilized syntax by default.

Given that https://github.com/facebook/jest/pull/4519 landed in Jest, it seems like this doesn't have any reason to be in Istanbul though.

Reply:
@SimenB can you verify this?

@loganfsmyth is there any reason we shouldn't be upgrading Istanbul to babel 7 as soon as possible? would this issue not arise in 7.

Reply:
> is there any reason we shouldn't be upgrading Istanbul to babel 7 as soon as possible?

It's probably good to update the plugin to support Babel 7 stuff, but we've done our best to make plugins work on both 6 and 7, so I think plugin should be careful not to drop support for 6 too fast.

> would this issue not arise in 7.

The newer version of Babel is likely to support the newer standards certainly, so I'd expect it to have object spread on by default, but there will always be more stuff.

I'll also say that if https://github.com/babel/babel/issues/7660 does become a thing, I'm not at all against Istanbul using it, but it seems like you'd want to be enabling it in [`instrumentSync`](https://github.com/istanbuljs/istanbuljs/blob/67918e2261a80acfafc0400b3980fafb6ae4aca5/packages/istanbul-lib-instrument/src/instrumenter.js#L77) rather than as part of the Babel plugin itself.

You've already got `objectRestSpread` hard-coded in https://github.com/istanbuljs/istanbuljs/blob/67918e2261a80acfafc0400b3980fafb6ae4aca5/packages/istanbul-lib-instrument/src/instrumenter.js#L89. If you changed that logic to use babel-core directly, it seems like things would be a lot simpler.


As a general overview of the current logic here and in `instrument`, I think there's a few things that seem surprising to me that complicate matters. Primarily, I think the current fact that the logic for the plugin doesn't actually live in the plugin leads to a lot of additional complexity. Mostly in the way that `istanbul-lib-instrument` directly depends on `babel-generate` and `babylon`, which are entirely unused via `babel-plugin-istanbul`. Similarly, `babel-types` is explicitly imported for Babel 6 here: https://github.com/istanbuljs/istanbuljs/blob/67918e2261a80acfafc0400b3980fafb6ae4aca5/packages/istanbul-lib-instrument/src/instrumenter.js#L6 but it is passed into the plugin here: https://github.com/istanbuljs/babel-plugin-istanbul/blob/master/src/index.js#L53 meaning that really the instrumentation logic should be using the version passed to the plugin, not importing it manually and locking yourselves to a specific version.

It seems like an ideal world we'd have

* `babel-plugin-istanbul` as a standard Babel plugin that worked on Babel 6 and Babel 7 and contained all of the logic for injecting coverage tracking.
* `istanbul-lib-instrument` which would wrap `@babel/core@7.x` and call it automatically with only `babel-plugin-istanbul` (potentially taking the user's local syntax plugins into account when parsing?)